### DISCUSSION TITLE
[Feature] Source generator for DependencyProperty boilerplate

### DISCUSSION THREAD

**Discussion by danielchalmers** [CommunityToolkit/Labs-Windows#593] [Ideas]
## Describe the problem this feature would solve

Like the [new source generators](https://github.com/CommunityToolkit/dotnet/issues/8) in `Microsoft.Toolkit.Mvvm`, we could have an attribute that generates the boilerplate for a `DependencyProperty`.

You can replace:

```csharp
private string foo;

public string Foo
{
    get => foo;
    set
    {
        foo = value;
        OnPropertyChanged(nameof(Foo));
    }
}
```

with

```csharp
[ObservableProperty]
public string Foo { get; set; }
```

With this, you could replace

```csharp
public static readonly DependencyProperty AquariumGraphicProperty = DependencyProperty.Register(
  "AquariumGraphic",
  typeof(Uri),
  typeof(AquariumObject)
);

public Uri AquariumGraphic
{
  get { return (Uri)GetValue(AquariumGraphicProperty); }
  set { SetValue(AquariumGraphicProperty, value); }
}
```

with

```csharp
[DependencyProperty]
public Uri AquariumGraphic { get; set; }
```

which is a huge reduction in code!

## Describe the solution

New feature that generates the boilerplate code for you.

## Describe alternatives you've considered

Snippets, copy paste.

## Additional context & Screenshots

- https://devblogs.microsoft.com/ifdef-windows/announcing-net-community-toolkit-v8-0-0-preview-1/
- https://docs.microsoft.com/en-us/dotnet/api/system.windows.dependencyproperty
- https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/custom-dependency-properties

**Comment by ghost** [comment]
Thanks for submitting a new feature request! I've automatically added a vote 👍 reaction to help get things started. Other community members can vote to help us prioritize this feature in the future!

**Comment by ghost** [comment]
Hello danielchalmers, thank you for opening an issue with us! 

I have automatically added a "needs triage" label to help get things started. Our team will analyze and investigate the issue, and escalate it to the relevant team if possible. Other community members may also look into the issue and provide feedback 🙌

**Comment by jader1313** [comment]
I believe it would be easy to implement and VERY VERY useful

**Comment by jader1313** [comment]
UNO Platform already has something implemented: 
https://platform.uno/docs/articles/uno-development/Internal-DependencyProperty-Generator.html

**Comment by chrisk414** [comment]
DependencyProperty requires rather ugly boilerplates, if not the ugliest, and it would be really nice to see it's supported.
It's doable, but unfortunately, MAUI decided to call it differently. Even though it's different, I believe the code generator should be able to tell the difference between WPF and MAUI when generating codes, no?

Anyway, I'm fully behind this feature and I hope the code generator's functionalities are expanded gradually.  It seems to slow down these days and I hope I'm wrong. 
Cheers!

**Comment by michael-hawker** [comment]
> [DependencyProperty]
> public Uri AquariumGraphic { get; set; }

The problem is this syntax doesn't exist yet in C#. That's the `partial properties` ([Discussion](https://github.com/dotnet/csharplang/discussions/3412)/[Issue](https://github.com/dotnet/csharplang/issues/6420)) feature which is really needed for this to work in C#.

Once that's done it'd be something to look at more closely, along with when [null conditional-assignment](https://github.com/dotnet/csharplang/issues/6045) becomes a thing, to help really clean-up our codebase around controls and properties.

For now, we don't have plans to do something like this until partial properties become a thing, but it's something that's very much of interest to us in order to reduce a ton of boilerplate in the Toolkit. 

For now, I'd suggest checking out this project: https://github.com/HavenDV/DependencyPropertyGenerator

**Reply by danielchalmers** [reply]
Hi @michael-hawker, is this back in focus with [C# 13](https://devblogs.microsoft.com/dotnet/csharp-13-explore-preview-features/#partial-properties)?

**Reply by michael-hawker** [reply]
@danielchalmers yes, @Sergio0694 is planning to work on something in the near future I believe. Just waiting to find out what else is making C# 13 or not.

We formalized a proposal for how the generators should behave over here in the Labs repo: https://github.com/CommunityToolkit/Labs-Windows/discussions/449

**Comment by DavidHollman** [comment]
FYI in https://github.com/HavenDV/DependencyPropertyGenerator they have implemented this as an attribute on the class.

**Comment by Arlodotexe** [comment]
Closing as duplicate of https://github.com/CommunityToolkit/Labs-Windows/discussions/449
